cmd/tpm-manager/tpm-manager.go

```diff
@@ -0,0 +1,507 @@
+	newstate := isTrusted(node)
+	newnode, err := manager.client.Nodes().Get(node.Name)
+	if err != nil {
+		glog.Errorf("Unable to obtain node state for %s: %v", node.Name, err)
+		return
+	}
+	if newnode.ObjectMeta.Annotations == nil {
+		newnode.ObjectMeta.Annotations = make(map[string]string)
+	}
+	currenttime := time.Now().Unix()
+
+	// If we're transitioning state, update the metadata
+	if isTrusted(newnode) != newstate {
+		if newstate == false {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = strconv.FormatInt(currenttime, 10)
+			newnode.ObjectMeta.Annotations[TrustedSince] = ""
+		} else {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = ""
+			newnode.ObjectMeta.Annotations[TrustedSince] = strconv.FormatInt(currenttime, 10)
+		}
+	}
+
+	newnode.ObjectMeta.Annotations[ValidationTime] = strconv.FormatInt(currenttime, 10)
+
+	// Ensure that the new node is tainted appropriately
+	if newstate == true {
+		trustNode(newnode)
+	} else {
+		invalidateNode(newnode)
+	}
+	newnode.ObjectMeta.Annotations[LogState] = node.ObjectMeta.Annotations[LogState]
+	newnode, err = manager.client.Nodes().Update(newnode)
+	if err != nil {
+		glog.Errorf("Unable to update node state for %s: %v", node.Name, err)
+		return
+	}
+}
+
+func verifyAllNodes() {
+	nodes, err := manager.client.Nodes().List(api.ListOptions{})
+	if err != nil {
+		glog.Errorf("Unable to obtain list of nodes")
+		return
+	}
+	for _, node := range nodes.Items {
+		verifyAndUpdate(&node)
+	}
+}
+
+func reverify() {
+	for {
+		select {
+		case <-time.After(time.Duration(manager.recurring) * time.Second):
+			verifyAllNodes()
+		case <-manager.recurringChan:
+		}
+	}
+}
+
+type TPMManager struct {
+	Master        string
+	Kubeconfig    string
+	tpmhandler    tpm.TPMHandler
+	pcrConfig     string
+	pcrConfigDir  string
+	allowUnknown  bool
+	recurring     int
+	client        *client.Client
+	policyTimer   *time.Timer
+	leaderelect   componentconfig.LeaderElectionConfiguration
+	recurringChan chan int
+	pcrConfigs    []map[string]tpm.PCRConfig
+}
+
+var manager TPMManager
+
+func addFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&manager.Master, "master", manager.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig)")
+	fs.StringVar(&manager.Kubeconfig, "kubeconfig", manager.Kubeconfig, "Path to kubeconfig file with authorization and master location information.")
+	fs.StringVar(&manager.pcrConfig, "pcrConfig", manager.pcrConfig, "Path to a single PCR config file")
+	fs.StringVar(&manager.pcrConfigDir, "pcrConfigDir", manager.pcrConfigDir, "Path to a PCR config directory")
+	fs.BoolVar(&manager.allowUnknown, "allowUnknown", false, "Allow unknown TPMs to join the cluster")
+	fs.IntVar(&manager.recurring, "reverify", 0, "Periodocally reverify nodes after this many seconds")
+}
+
+func updateConfig(configmap *api.ConfigMap) {
+	if configmap.Data["allowunknown"] != "" {
+		allowUnknown, err := strconv.ParseBool(configmap.Data["allowunknown"])
```

Comment:
ignoring errors on subtle things like failing to parse an incorrect configMap field worries me. What if an admin mistypes the value?